core/API/CORSHandler.php

```diff
@@ -18,40 +20,85 @@ class CORSHandler
      */
     protected $domains;
 
+    private $logger;
+
     public function __construct()
     {
         $this->domains = Url::getCorsHostsFromConfig();
+        $this->logger = StaticContainer::get(LoggerInterface::class);
+
     }
 
+    /**
+     * This method is set header for the request to pass browser cross domain checks.
+     * By default, it allow from all hosts
+     * @throws \Exception
+     */
     public function handle()
     {
-        if (empty($this->domains)) {
-            return;
-        }
-        
+
+        // set default header
         Common::sendHeader('Vary: Origin');
-        
-        // allow Piwik to serve data to all domains
-        if (in_array("*", $this->domains)) {
-            
-            Common::sendHeader('Access-Control-Allow-Credentials: true');
-            
-            if (!empty($_SERVER['HTTP_ORIGIN'])) {
-                Common::sendHeader('Access-Control-Allow-Origin: ' . $_SERVER['HTTP_ORIGIN']);
-                return;
-            }
-            
-            Common::sendHeader('Access-Control-Allow-Origin: *');
-            return;
-        }
+        Common::sendHeader('Access-Control-Allow-Credentials: true');
+        Common::sendHeader('Access-Control-Allow-Origin: *');
 
-        // specifically allow if it is one of the allowlisted CORS domains
+
+        // when origin is set, response http origin as response
         if (!empty($_SERVER['HTTP_ORIGIN'])) {
-            $origin = $_SERVER['HTTP_ORIGIN'];
-            if (in_array($origin, $this->domains, true)) {
-                Common::sendHeader('Access-Control-Allow-Credentials: true');
-                Common::sendHeader('Access-Control-Allow-Origin: ' . $_SERVER['HTTP_ORIGIN']);
+            Common::sendHeader('Access-Control-Allow-Origin: ' . $_SERVER['HTTP_ORIGIN']);
+        }
+
+        //check if http origin is not in the cor_domain list
+        if (!empty($this->domains) && !empty($_SERVER['HTTP_ORIGIN'])) {
+            if (!in_array('*', $this->domains) && !in_array($_SERVER['HTTP_ORIGIN'], $this->domains, true)) {
+                Common::sendHeader('Access-Control-Allow-Origin: ' . $this->domains[0], true);
+                Common::sendResponseCode(401);
+                $this->logger->debug("Tracker detected CORS request. Skipping...");
```

Comment:
Actually the CORS class is not only used in Tracker, so this log (and the one below) might be incorrect, by saying the Tracker detected.